madsim/src/sim/net/tcp/network.rs

```diff
@@ -0,0 +1,267 @@
+        let mut inner = self.inner.lock().unwrap();
+        f(&mut inner.config);
+    }
+
+    pub(crate) fn reset_node(&self, node: NodeId) -> Result<(), String> {
+        trace!("reset node {}", node);
+        let mut inner = self.inner.lock().unwrap();
+        let mut drop_ids = vec![];
+
+        for (conn_id, conn) in &inner.conn {
+            if conn.src == node || conn.dst == node {
+                drop_ids.push(*conn_id);
+            }
+        }
+
+        if drop_ids.is_empty() {
+            Err("there is no node to reset".to_string())
+        } else {
+            for id in drop_ids {
+                inner.conn.remove(&id);
+            }
+
+            Ok(())
+        }
+    }
+
+    pub(crate) fn drop_connection(&self, conn: ConnId) -> io::Result<()> {
+        let mut inner = self.inner.lock().unwrap();
+        let removed = inner.conn.remove(&conn).is_some();
+        if removed {
+            Ok(())
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::NotConnected,
+                "connection not exist or already shutdown".to_string(),
+            ))
+        }
+    }
+
+    pub(crate) fn clog_node(&self, id: NodeId) {
+        trace!("tcp clog: {id}");
+        self.inner.lock().unwrap().clogged_node.insert(id);
+    }
+
+    pub(crate) fn unclog_node(&self, id: NodeId) {
+        trace!("tcp unclog: {id}");
+        self.inner.lock().unwrap().clogged_node.remove(&id);
+    }
+
+    pub(crate) fn clog_link(&self, src: NodeId, dst: NodeId) {
+        trace!("clog: {src} -> {dst}");
+        self.inner.lock().unwrap().clogged_link.insert((src, dst));
+    }
+
+    pub(crate) fn unclog_link(&self, src: NodeId, dst: NodeId) {
+        trace!("tcp unclog: {src} -> {dst}");
+        self.inner.lock().unwrap().clogged_link.remove(&(src, dst));
+    }
+
+    pub(crate) fn listen(
+        &self,
+        id: NodeId,
+        addr: SocketAddr,
+    ) -> Result<mpsc::Receiver<(ConnId, ConnId)>, String> {
+        let (tx, rx) = mpsc::channel(0);
+        let is_inuse = self
+            .inner
+            .lock()
+            .unwrap()
+            .accept
+            .insert(addr, (id, tx))
+            .is_some();
+        if is_inuse {
+            Err("addr is in used".to_string())
+        } else {
+            Ok(rx)
+        }
+    }
+
+    pub(crate) async fn connect(
+        &self,
+        src: NodeId,
+        dst: &SocketAddr,
+    ) -> Result<(ConnId, ConnId), String> {
+        let (send_conn, recv_conn, mut tx) = {
+            let mut inner = self.inner.lock().unwrap();
+
+            let (dst, tx) = if let Some((id, tx)) = inner.accept.get_mut(dst) {
+                (*id, tx.clone())
+            } else {
+                return Err("addr not be listened".to_string());
+            };
+
+            let send_conn = inner.conn_cnt.fetch_add(1, Ordering::SeqCst);
+            let recv_conn = inner.conn_cnt.fetch_add(1, Ordering::SeqCst);
```

Comment:
reduce atomic operations
```suggestion
            let send_conn = inner.conn_cnt.fetch_add(2, Ordering::SeqCst);
            let recv_conn = send_conn + 1;
```